server/scenario/web_api/data/action.py

```diff
@@ -0,0 +1,185 @@
+import json
+import logging
+
+from flask import Blueprint, request
+from werkzeug.exceptions import NotFound
+
+import models
+from app_config import db
+from scenario.web_api.utils import update_media
+from utils.decorator import RequiredValueSource, required
+from vars import RESULT_DELIMITER
+
+# pyright: reportCallIssue=false
+# pyright: reportAttributeAccessIssue=false
+# The following statements are excluded from pyright, due to ORM specifics.
+
+
+web_api = Blueprint("web_api-action", __name__)
+
+
+@web_api.get("/action/all")
+def get_all_actions():
+    """ Returns a json of all actions stored. """
+    action_list = models.Action.query.all()
+    return [
+        {
+            "id": action.id,
+            "name": action.name,
+        } for action in action_list
+    ]
+
+
+@web_api.get("/action")
+@required("action_id", int, RequiredValueSource.ARGS)
+def get_action(action_id: int):
+    """ Returns an action with all its related entries. """
+    action = models.Action.query.filter_by(id=action_id).first()
+
+    if not action:
+        raise NotFound(f"action not found by id={action_id}")
+
+    min_role = models.Role.query.filter_by(power=action.required_power).first()
+
+    resource_list = models.Resource.query.join(
+        models.ResourcesNeeded,
+        models.Resource.id == models.ResourcesNeeded.resource_id
+    ).filter(
+        models.ResourcesNeeded.action_id == action_id
+    ).all()
+
+    return {
+        "id": action.id,
+        "name": action.name,
+        "min_role": min_role.name if min_role else None,
+        "duration_secs": action.duration_secs,
+        "media_refs": json.loads(action.media_refs) if action.media_refs else [{}],
+        "results": action.results.split(RESULT_DELIMITER),
+        "resources": [
+            {
+                "id": resource.id,
+                "name": resource.name
+            } for resource in resource_list
+        ]
+    }
+
+
+@web_api.post("/action")
+def create_action():
+    """ Creates an empty Action and returns the dbo as JSON. """
+    action = models.Action(
+        name="Neue Maßnahmen",
+        duration_secs=0,
+        required_power=0,
+        media_refs=[{}],
+        result=""
+    )
+    db.session.add(action)
+    db.session.commit()
```

Comment:
What if you hit the new action button twice without editing the first or is this not possible?

Reply:
the action id autoincrements and you get 2 action with the same name displayed. no collision on db-side. further, you will be redirected in a different edit window, so you cant create 2 actions. Locations and resource behave the same